README.md
# OpenSCENARIO2SUMO
Automatic traffic scenario conversion interface between OpenSCENARIO and the traffic simulator SUMO. Currently, only the conversion from **O**pen**SC**ENARIO to **C**ommon**R**OAD (osc2cr) is developed.<br>

### Code Reference
This repo refers to 
1. [Commonroad Scenario Designer](https://gitlab.lrz.de/tum-cps/commonroad-scenario-designer.git)
2. [CommonROAD-OpenSCENARIO Converter](https://gitlab.lrz.de/tum-cps/commonroad-openscenario-converter.git)

### Requirements
The code is written in Python 3.10 and has been tested on Ubuntu 22.04. 

If you want to use the SUMO conversion or to generate traffic using SUMO, please install 
[SUMO](https://sumo.dlr.de/docs/index.html):
```bash
sudo apt-get update
sudo apt-get install sumo sumo-tools sumo-doc
echo "export SUMO_HOME=/usr/share/sumo" >> ~/.bashrc
echo 'export PYTHONPATH="$SUMO_HOME/tools:$PYTHONPATH"' >> ~/.bashrc
```
If you use zsh, replace `.bashrc` with `.zshrc`.


### Using the CommonRoad Scenario Designer
The recommended way of installation if you only want to use the scenario designer (i.e., you do not want to work with the code directly) is to use the PyPI package:
```bash
pip install commonroad-scenario-designer
```


### Common Errors during installation

- **Could not load the Qt platform plugin “xcb” in “” even though it was found:** Error seems to be a missing package - either libxkbcommon-x11 or libxcb-xinerama0 (both can be installed by ```sudo apt install [package_name]```). See for reference [here](https://discuss.pixls.us/t/solved-could-not-load-the-qt-platform-plugin-xcb-in-even-though-it-was-found/17677/9)




### GUI Usage

The GUI can either be activated via the command line via the following two options:

```bash
$ crdesigner
$ crdesigner gui
```
Note that you have to activate first the Python environment in which the CommonRoad Scenario Designer is installed.





### Openscenario to Commonroad Converter
We recommend using [Anaconda](https://www.anaconda.com/) to manage your environment so that
even if you mess something up, you can always have a safe and clean restart. 
A guide for managing Python environments with Anaconda can be found [here](https://conda.io/projects/conda/en/latest/user-guide/tasks/manage-environments.html).

- First, clone the repository. 
- After installing Anaconda, create a new environment with (>=3.9) and enter it:
``` bash
$ conda create -n commonroad-py39 python=3.9 -y
$ conda activate commonroad-py39
```
- Then, install the dependencies with:

```bash
$ pip install commonroad-openscenario-converter
```

- Test case:
```bash
$ cd Openscenario2Commonroad/test
$ python OSC_Convert2_CR.py
```
The generated simple curve scenario "OpenSCENRAIO2SUMO/output/SimpleCurve/OSC_SimpleCurve_1_T-1.xml" could be opened via the CommonRoad Scenario Designer.

![pics/src2cr.png](https://github.com/TianZheng095/OpenSCENARIO2SUMO/blob/main/pics/src2cr.png)

### Open Commonroad file with crdesigner GUI 
Open the generated .xml file using the crdesigner GUI

![pics/openfileGUI.png](https://github.com/TianZheng095/OpenSCENARIO2SUMO/blob/main/pics/openfileGUI)

There are some redundant roads generated from Openscenario to Commonroad. They are able to be removed in this GUI to achieve a better and more reasonable conversion to SUMO later.
![pics/circle.png](https://github.com/TianZheng095/OpenSCENARIO2SUMO/blob/main/pics/circle.png)


### Commonroad to SUMO conversion
Fist way: using crdesigner GUI

![pics/coversion2SUMO.png](https://github.com/TianZheng095/OpenSCENARIO2SUMO/blob/main/pics/coversion2SUMO.png)

- Advantage: easy to operate
- Shortcoming: no clear error message when encountering an error

Second way: using the script
```bash
$ cd Commonroad2SUMO/tutorials/conversion_examples
$ python example_commonroad_to_sumo.py
```


- Advantage: clear error message when encountering an error
- Shortcoming: complicate to operate

![pics/sumo.png](https://github.com/TianZheng095/OpenSCENARIO2SUMO/blob/main/pics/sumo.png)

### Paper Reference
1. [CommonRoad Scenario Designer: An Open-Source Toolbox for Map Conversion and Scenario Creation for Autonomous Vehicles](https://arxiv.org/pdf/2305.10080.pdf)
2. [Automatic Traffic Scenario Conversion from OpenSCENARIO to CommonRoad](https://arxiv.org/pdf/2305.10080.pdf)


Commonroad2SUMO/CHANGELOG.md
# Changelog

## [0.7.0] - 2023-03-XX

### Added
- Possibility to load aerial image from Bing in the background
- Possibility to load aerial image from LDBV in the background
- CommonRoad to Lanelet2 conversion: Conversion of regulatory elements
- CommonRoad to Lanelet2 conversion: Conversion of lanelet types

### Changed
- Default matplotlib area
- Generalized traffic sign conversion 
- Lanelet2 to CommonRoad conversion: Generalize traffic sign conversion
- Packaging using poetry

### Fixed
- Persisting zoomed scale
- Connect to predecessor, previously added, and successor
- Adding rotated lanelet (>360°)
- Lanelet selection after translating a lanelet in the GUI
- Blue position waypoint in GUI is removed when new scenario is created
- Video and matplotlib figure saving not working

### Removed

## [0.6.1] - 2023-02-09

### Fixed
- GUI setting for axis visibility not changeable
- OpenDRIVE conversion endless loop speed limit mapping
- Yaml configuration loading under Windows

## [0.6.0] - 2023-01-31

### Added
- New test cases for the OpenDRIVE to CommonRoad conversion
- New shortcut buttons for lanelet adding, merging, splitting, creating adjacent
- Function to convert OSM to CommonRoad using SUMO
- Convert lane speed limits in OpenDRIVE
- Convert OpenDRIVE stop lines represented as object
- Reading of protobuf CommonRoad scenarios
- Support for commonroad-io 2022.3
- Unit test cases for Lanelet2 conversion
- Removing lanelets via "entf", "back", and "del" keys
- Removing lanelets via right click on lanelet
- Open edit section of lanelet via right click on lanelet
- Support for Python 3.10
- Preserving of settings

### Changed
- User interface for adding and updating lanelets

### Fixed
- Preventing GUI crash when adding obstacle without existing scenario
- Conversion of straight euler spiral 
- Floating point error in computation of Cartesian points of lane border in OpenDRIVE2CR conversion
- Various small bug fixes

### Removed
- Support for Python 3.7

## [0.5.1] - 2022-05-20

### Fixed
- Switching shape of dynamic obstacle
- Adding adjacent lanelet which already exists
- Lanelet translation with missing x- or y-coordinate
- Missing location information when storing a scenario
- Various small bug fixes

## [0.5] - 2022-03-06

### Added
- Dynamic loading of large maps
- Adding of static obstacles
- Individual colors for obstacles
- Obstacle states can be changed by moving profiles in obstacle toolbox

### Fixed
- System crash if SUMO is available
- Undo button works again
- Various small bug fixes

## [0.4] - 2022-01-06

### Added
- Adding/Updating of static obstacles with circle, polygon, and rectangle shape in GUI

### Changed
- New GUI software architecture

### Fixed
- Various small bug fixes

## [0.3] - 2021-12-02

### Changed
- Solved deprecated warnings related to commonroad-io version 2021.3

### Fixed
- Various small bug fixes

## [0.2] - 2021-09-22

### Added
- Planning problem visualization in GUI

### Changed
- Extension of readme and documentation

### Fixed
- Various small bug fixes

## [0.1] - 2021-04-01
### Added
- graphical user interface for creating and manipulating CommonRoad maps and scenarios
- converts between different map formats: OpenStreetMap, SUMO, Lanelet/Lanelet2, OpenDRIVE, CommonRoad
- test cases for different converters
- tutorials for different converters
- readme and documentation


Commonroad2SUMO/crdesigner/map_conversion/osm2cr/config.py
"""
This module holds all parameters necessary for the conversion
"""

# Benchmark Settings
# name of the benchmark
BENCHMARK_ID = "ZAM_Test-1_1_T-1"
# author of the benchmark
AUTHOR = "Automated converter by Maximilian Rieger"
# affiliation of the benchmark
AFFILIATION = "Technical University of Munich, Germany"
# source of the benchmark
SOURCE = "OpenStreetMaps (OSM)"
# additional tags for the benchmark
TAGS = "urban"
# Geonames username to retrieve geonamesID for created scenarios
GEONAMES_USERNAME = "demo"
# Mapillary Client ID which can be set to extract additional traffic signs. If set to "demo",
# mapillary signs will be disabled
MAPILLARY_CLIENT_ID = "demo"
# time step size for the benchmark in seconds
TIMESTEPSIZE = 0.1

# Lanelet type each lanelet should have
LANELETTYPE = "urban"

# Aerial Image Settings
# Use aerial images for edit
AERIAL_IMAGES = False
# Path to save downloaded aerial images
IMAGE_SAVE_PATH = "files/imagery/"
# The zoom level of Bing Maps tiles
ZOOM_LEVEL = 19
# aerial image area threshold limiting the user input for the coordinates
AERIAL_IMAGE_THRESHOLD = 0.01

# Map download Settings
# path to save downloaded files
SAVE_PATH = "files/"
# half width of area downloaded in meters
DOWNLOAD_EDGE_LENGTH = 200
# coordinates in latitude and longitude specifying the center of the downloaded area
DOWNLOAD_COORDINATES = (48.262447, 11.657881)

# Scenario Settings
# include tunnels in result
LOAD_TUNNELS = False
# delete unconnected edges
MAKE_CONTIGUOUS = False
# split edges at corners (~90° between two waypoint segments)
# this can help to model the course of roads on parking lots better
SPLIT_AT_CORNER = True
# use OSM restrictions for linking process
USE_RESTRICTIONS = True
# types of roads extracted from the OSM file
# suitable types: 'motorway', 'trunk', 'primary', 'secondary', 'tertiary', 'unclassified', 'residential',
# 'motorway_link', 'trunk_link', 'primary_link', 'secondary_link', 'tertiary_link', 'living_street', 'service'
ACCEPTED_HIGHWAYS_MAINLAYER = [
    "motorway",
    "trunk",
    "primary",
    "secondary",
    "tertiary",
    # "unclassified",
    "residential",
    "motorway_link",
    "trunk_link",
    "primary_link",
    "secondary_link",
    "tertiary_link",
    "living_street",
    # "service"
]
EXTRACT_SUBLAYER = False
# types of highways extracted from the OSM file as sublayer
# elements mustn't be in ACCEPTED_HIGHWAYS
ACCEPTED_HIGHWAYS_SUBLAYER = [
    "path",
    "footway",
    "cycleway"
]
# Lanelet type of the sublayer lanelets
SUBLAYER_LANELETTYPE = 'sidewalk'
# Lanelet type of the sublayer lanelets that cross the main layer
# overwrites SUBLAYER_LANELETTYPE for lanelets applied on
CROSSING_LANELETTYPE = 'crosswalk'
# osm ways with these tags are not taken into account
REJECTED_TAGS = {
    "area": "yes"
}
# number of lanes for each type of road should be >=1
LANECOUNTS = {
    "motorway": 6,
    "trunk": 4,
    "primary": 2,
    "secondary": 2,
    "tertiary": 2,
    "unclassified": 2,
    "residential": 2,
    "motorway_link": 2,
    "trunk_link": 2,
    "primary_link": 2,
    "secondary_link": 2,
    "tertiary_link": 2,
    "living_street": 2,
    "service": 2,
    "path": 1,
    "footway": 1,
    "cycleway": 1
}
# width of lanes for each type of road in meters
LANEWIDTHS = {
    "motorway": 3.5,
    "trunk": 3.5,
    "primary": 3.5,
    "secondary": 3.5,
    "tertiary": 3.5,
    "unclassified": 3.5,
    "residential": 3.5,
    "motorway_link": 3.5,
    "trunk_link": 3.5,
    "primary_link": 3.5,
    "secondary_link": 3.5,
    "tertiary_link": 3.5,
    "living_street": 3.5,
    "service": 3.5,
    "path": 2.0,
    "footway": 2.0,
    "cycleway": 2.0
}
# default speed limit for each type of road in km/h
SPEED_LIMITS = {
    "motorway": 120,
    "trunk": 100,
    "primary": 100,
    "secondary": 100,
    "tertiary": 100,
    "unclassified": 80,
    "residential": 50,
    "motorway_link": 80,
    "trunk_link": 80,
    "primary_link": 80,
    "secondary_link": 80,
    "tertiary_link": 80,
    "living_street": 7,
    "service": 10,
    "path": 8,
    "footway": 8,
    "cycleway": 20
}

# Export Settings
# desired distance between interpolated waypoints in meters
INTERPOLATION_DISTANCE = 0.5
# allowed inaccuracy of exported lines to reduce number of way points in meters
COMPRESSION_THRESHOLD = 0.05
# export the scenario in UTM coordinates
EXPORT_IN_UTM = False
# toggle filtering of negligible waypoints
FILTER = True
# delete invalid lanes before export
DELETE_INVALID_LANES = True

# Internal Settings (these can be used to improve the conversion process for individual scenarios)
# radius of the earth used for calculation in meters
EARTH_RADIUS = 6371000
# delete short edges after cropping
DELETE_SHORT_EDGES = False
# distance between waypoints used internally in meters
INTERPOLATION_DISTANCE_INTERNAL = 0.5
# bezier parameter for interpolation (should be within [0, 0.5])
BEZIER_PARAMETER = 0.35
# distance between roads at intersection used for cropping in meters
INTERSECTION_DISTANCE = 4.0
# associated with pedestrian pathways by default
INTERSECTION_DISTANCE_SUBLAYER = 1.0
# defines if the distance to other roads is used for cropping
# if false the distance to the center of the intersection is used
INTERSECTION_CROPPING_WITH_RESPECT_TO_ROADS = True
# threshold above which angles are considered as soft in degrees
SOFT_ANGLE_THRESHOLD = 55.0
# least angle for lane segment to be added to the graph in degrees.
# if you edit the graph by hand, a value of 0 is recommended
LANE_SEGMENT_ANGLE = 5.0
# least distance between graph nodes to try clustering in meters
CLUSTER_LENGTH = 10.0
# least length of cluster to be added in meters
LEAST_CLUSTER_LENGTH = 10.0
# maximal distance between two intersections to which they are merged, if zero, no intersections are merged
MERGE_DISTANCE = 3.5
# threshold which is used to determine if a successor of an incoming lane is considered as straight
INTERSECTION_STRAIGHT_THRESHOLD = 35.0
# option to clean up intersections and add new traffic lights to it, that are not part of the original OSM file
INTERSECTION_ENHANCEMENT = True
# option to remove unconnected lanelets from the main lanelet scenario
REMOVE_UNCONNECTED_LANELETS = True
# set of processed turn lanes
# this should only be changed for further development
RECOGNIZED_TURNLANES = [
    "left",
    "through",
    "right",
    "slight_right",
    "slight_left",
    "merge_to_left",
    "merge_to_right",
    "through;right",
    "left;through",
    "left;through;right",
    "left;right",
    "none",
]

# Traffic Lights
# cycle that will be applied to each traffic light. Values in seconds
TRAFFIC_LIGHT_CYCLE = {
    "red_phase": 57,
    "red_yellow_phase": 3,
    "green_phase": 37,
    "yellow_phase": 3
}

# Traffic Signs
# values to search for in OSM
TRAFFIC_SIGN_VALUES = [
    "traffic_signals",
    "stop",
    "give_way",
    "city_limit",
]
# keys to search for in OSM
TRAFFIC_SIGN_KEYS = [
    "traffic_sign",
    "overtaking",
    "traffic_signals:direction",
    "maxspeed",
]
# categories to include if mapillary is used for sign extraction
MAPILLARY_CATEGORIES = [
    "warning",
    "regulatory",
    "information",
    "complementary"
]
# include traffic signs based on their id, e.g. "Max_SPEED". Keep "ALL" to accept all found traffic sings
ACCEPTED_TRAFFIC_SIGNS = ["ALL"]
# exclude traffic signs based on their id, e.g. "MAX_SPEED". "ALL" has to be set in ACCEPTED_TRAFFIC_SIGNS
EXCLUDED_TRAFFIC_SIGNS = []

# User edit activation
# Toggle edit for user
USER_EDIT = False
